### PR TITLE
Normalize type parameters in create_mono_items_for_default_impls.

### PR DESCRIPTION
**Pull request by alessandrod**
Fix for https://github.com/rust-lang/rust/issues/58375. I've added a test in `src/tests/run-pass` to reproduce the bug, not sure that's the best place for it.

See https://github.com/rust-lang/rust/issues/58375#issuecomment-509156977 for more context.

### CODE REVIEW

**Comment by varkor** [comment]
r? @eddyb 

This looks reasonable, but I'm not confident knowing when normalisation is appropriate, and don't want to overlook something.

File: src/librustc_mir/monomorphize/collector.rs [hunk 1]
```diff
@@ -1166,8 +1166,13 @@ fn create_mono_items_for_default_impls<'tcx>(
                         }
                     });
 
+                    let param_env = ty::ParamEnv::reveal_all();
+                    let substs = tcx.normalize_erasing_regions(
+                        param_env,
+                        substs,
+                    );
```
Comment (eddyb) [inline src/librustc_mir/monomorphize/collector.rs]: You should be able to do this on the `trait_ref`, I think? The problem appears to be that `impl_trait_ref` might return an unnormalized `TraitRef`, the per-method substs just copy from there.

**Comment by alessandrod** [comment]
Thank you for reviewing this so quickly @varkor and @eddyb 🙏

**Pull request merged**